File: Sources/SimConsole/SimConsole.swift
import Foundation
import os

/// Entry point for the in-app side of `sim-console`.
///
/// Routes structured events through `os.Logger` so the companion `sim-console`
/// macOS app — which subscribes to `xcrun simctl spawn <UDID> log stream` —
/// can pick them up via predicates that match `subsystem == <bundle-id> AND
/// category == "analytics" | "network" | "event"`.
///
/// All payloads are emitted with `%{public}@` so the unified logging system
/// won't redact them to `<private>`. Use in **debug builds only** — never ship
/// this enabled in a production binary.
///
/// Wire by calling `SimConsole.bootstrap(...)` once at launch, e.g.:
///
///     #if DEBUG
///     SimConsole.bootstrap(.init(subsystem: Bundle.main.bundleIdentifier ?? "app"))
///     #endif
public enum SimConsole {

    public struct Configuration {
        public var subsystem: String
        public var enabled: Bool
        public var maxBodyChars: Int

        public init(
            subsystem: String,
            enabled: Bool = true,
            maxBodyChars: Int = 800
        ) {
            self.subsystem = subsystem
            self.enabled = enabled
            self.maxBodyChars = maxBodyChars
        }
    }

    private static var configuration: Configuration?
    private static let queue = DispatchQueue(label: "com.simconsole.emit")

    private static var analyticsLogger: Logger?
    private static var networkLogger: Logger?
    private static var eventLogger: Logger?

    public static func bootstrap(_ config: Configuration) {
        queue.sync {
            configuration = config
            guard config.enabled else { return }
            analyticsLogger = Logger(subsystem: config.subsystem, category: "analytics")
            networkLogger   = Logger(subsystem: config.subsystem, category: "network")
            eventLogger     = Logger(subsystem: config.subsystem, category: "event")
        }
        // MockStore reads ~/.sim-console/mocks-<bundle>.json. Inert outside
        // the simulator. Wiring it up here means SimConsoleURLProtocol can
        // consult it on every outbound request without further setup.
        MockStore.shared.configure(bundleId: config.subsystem)
    }

    public static var isEnabled: Bool {
        queue.sync { configuration?.enabled == true }
    }

    static var maxBodyChars: Int {
        queue.sync { configuration?.maxBodyChars ?? 800 }
    }

    // MARK: Analytics

    public static func analytics(event: String, params: [String: Any] = [:], screen: String? = nil) {
        guard let logger = analyticsLogger else { return }
        let payload: [String: Any] = [
            "kind": "analytics",
            "t": now(),
            "event": event,
            "params": jsonSafe(params),
            "screen": screen as Any
        ]
        logger.log("\(json(payload), privacy: .public)")
    }

    public static func screen(_ name: String, params: [String: Any] = [:]) {
        guard let logger = analyticsLogger else { return }
        let payload: [String: Any] = [
            "kind": "screen",
            "t": now(),
            "screen": name,
            "params": jsonSafe(params)
        ]
        logger.log("\(json(payload), privacy: .public)")
    }

    // MARK: Network

    public static func network(request id: String, method: String, url: String, headers: [String: String] = [:], body: String? = nil, mocked: Bool = false) {
        guard let logger = networkLogger else { return }
        var payload: [String: Any] = [
            "kind": "net.request",
            "t": now(),
            "id": id,
            "method": method,
            "url": url
        ]
        if mocked { payload["mocked"] = true }
        logger.log("\(json(payload), privacy: .public)")
        emitHeaders(id: id, direction: "request", headers: headers)
        emitBody(id: id, direction: "request", body: body)
    }

    public static func network(response id: String, status: Int, durationMs: Int, headers: [String: String] = [:], body: String? = nil, byteSize: Int? = nil, mocked: Bool = false) {
        guard let logger = networkLogger else { return }
        var payload: [String: Any] = [
            "kind": "net.response",
            "t": now(),
            "id": id,
            "status": status,
            "duration_ms": durationMs
        ]
        if let byteSize { payload["byte_size"] = byteSize }
        if mocked { payload["mocked"] = true }
        logger.log("\(json(payload), privacy: .public)")
        emitHeaders(id: id, direction: "response", headers: headers)
        emitBody(id: id, direction: "response", body: body)
    }

    /// Each header is emitted as its own log line. A single bloated CDN header
    /// (report-to, nel, alt-svc, etc.) can be ~300 chars on its own — bundled
    /// into one payload they overflow `os_log`'s per-line limit and the whole
    /// event gets truncated before the parser can see `"kind"`. Per-header
    /// events trade volume for reliability: each line is small and parseable.
    private static func emitHeaders(id: String, direction: String, headers: [String: String]) {
        guard let logger = networkLogger, !headers.isEmpty else { return }
        for (name, value) in headers {
            let payload: [String: Any] = [
                "kind": "net.header",
                "t": now(),
                "id": id,
                "direction": direction,
                "name": name,
                "value": clip(value) ?? ""
            ]
            logger.log("\(json(payload), privacy: .public)")
        }
    }

    /// Body events live on their own log line so the meta row above can't be
    /// truncated by `os_log`'s per-line limit. The companion macOS app
    /// pairs them to the meta row by `id` + `direction`.
    private static func emitBody(id: String, direction: String, body: String?) {
        guard let logger = networkLogger,
              let body, !body.isEmpty else { return }
        let payload: [String: Any] = [
            "kind": "net.body",
            "t": now(),
            "id": id,
            "direction": direction,
            "body": clip(body) as Any
        ]
        logger.log("\(json(payload), privacy: .public)")
    }

    public static func network(error id: String, durationMs: Int, error: String) {
        guard let logger = networkLogger else { return }
        let payload: [String: Any] = [
            "kind": "net.error",
            "t": now(),
            "id": id,
            "duration_ms": durationMs,
            "error": error
        ]
        logger.log("\(json(payload), privacy: .public)")
    }

    // MARK: Generic event

    public enum Level: String {
        case debug, info, warn, error
    }

    public static func log(_ message: String, level: Level = .info, fields: [String: Any] = [:]) {
        guard let logger = eventLogger else { return }
        let payload: [String: Any] = [
            "kind": "log",
            "t": now(),
            "level": level.rawValue,
            "msg": message,
            "fields": jsonSafe(fields)
        ]
        switch level {
        case .debug: logger.debug("\(json(payload), privacy: .public)")
        case .info:  logger.info("\(json(payload), privacy: .public)")
        case .warn:  logger.warning("\(json(payload), privacy: .public)")
        case .error: logger.error("\(json(payload), privacy: .public)")
        }
    }

    // MARK: Helpers

    private static func now() -> Double { Date().timeIntervalSince1970 }

    private static func clip(_ s: String?) -> String? {
        guard let s, !s.isEmpty else { return s }
        let max = maxBodyChars
        if s.count <= max { return s }
        let head = s.prefix(max)
        return "\(head)…[+\(s.count - max) chars]"
    }

    private static func jsonSafe(_ dict: [String: Any]) -> [String: Any] {
        var out: [String: Any] = [:]
        for (k, v) in dict {
            out[k] = sanitize(v)
        }
        return out
    }

    private static func sanitize(_ value: Any) -> Any {
        switch value {
        case let s as String: return s
        case let n as NSNumber: return n
        case let b as Bool: return b
        case let arr as [Any]: return arr.map(sanitize)
        case let dict as [String: Any]: return jsonSafe(dict)
        case is NSNull: return NSNull()
        default: return String(describing: value)
        }
    }

    static func json(_ payload: [String: Any]) -> String {
        guard JSONSerialization.isValidJSONObject(payload),
              let data = try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys, .withoutEscapingSlashes]),
              let s = String(data: data, encoding: .utf8)
        else {
            return "{\"kind\":\"log\",\"msg\":\"sim-console: unencodable payload\"}"
        }
        return s
    }
}

File: android/simconsole/src/main/kotlin/com/simconsole/SimConsole.kt
package com.simconsole

import android.content.Context
import android.util.Log
import java.util.UUID

/**
 * Entry point for the in-app side of `sim-console` on Android.
 *
 * Routes structured events through `android.util.Log` with stable tags so the
 * companion `sim-console` macOS app — which subscribes to `adb logcat -v threadtime`
 * with a `SimConsole.*:V *:S` filter — can pick them up.
 *
 * The JSON envelope is byte-for-byte compatible with the iOS SDK so the panel
 * parses both platforms with one code path. See [Envelope] for the exact shape.
 *
 * **Debug-only.** Calls become no-ops if `bootstrap` was never invoked or if it
 * was called with `enabled = false`. Don't ship this enabled in a production binary.
 *
 * Typical wiring in an `Application.onCreate()`:
 *
 * ```kotlin
 * if (BuildConfig.DEBUG) {
 *     SimConsole.bootstrap(this, subsystem = BuildConfig.APPLICATION_ID)
 * }
 * ```
 */
object SimConsole {

    enum class Level { Debug, Info, Warn, Error }

    /**
     * Pluggable transport so unit tests can capture emissions without an emulator.
     * Production uses [LogcatSink], which calls into `android.util.Log`.
     */
    internal interface Sink {
        fun emit(tag: String, level: Level, payload: String)
    }

    data class Configuration(
        val subsystem: String,
        val enabled: Boolean = true,
        val maxBodyChars: Int = 800,
    )

    private val lock = Any()
    @Volatile private var configuration: Configuration? = null
    @Volatile internal var sink: Sink = LogcatSink

    private var metricSampler: MetricSampler? = null
    private var fpsSampler: FpsSampler? = null
    private var hangDetector: HangDetector? = null

    /** True iff [bootstrap] has been called with `enabled = true`. */
    val isEnabled: Boolean get() = configuration?.enabled == true

    internal val maxBodyChars: Int get() = configuration?.maxBodyChars ?: 800

    /**
     * Wire the SDK. Idempotent — repeated calls overwrite the config. The
     * `context.applicationContext` is retained for the lifetime of the process
     * so background samplers can query system services without leaking an
     * Activity reference.
     */
    @JvmStatic
    @JvmOverloads
    fun bootstrap(context: Context, subsystem: String, enabled: Boolean = true, maxBodyChars: Int = 800) {
        bootstrap(
            config = Configuration(subsystem = subsystem, enabled = enabled, maxBodyChars = maxBodyChars),
            appContext = context.applicationContext,
        )
    }

    @JvmStatic
    fun bootstrap(config: Configuration) {
        bootstrap(config = config, appContext = null)
    }

    private fun bootstrap(config: Configuration, appContext: Context?) {
        synchronized(lock) {
            configuration = config
            // Anchor launch timing immediately so milestones report from the
            // SDK-load moment if the host never calls Metric.appStartLaunch().
            Metric.processStart
            if (!config.enabled) {
                stopSamplersLocked()
                return
            }
            // Idempotent: don't stack samplers if bootstrap is called again.
            stopSamplersLocked()
            if (appContext != null) {
                metricSampler = MetricSampler(appContext).also { it.start() }
            }
            fpsSampler = FpsSampler().also { it.start() }
            hangDetector = HangDetector().also { it.start() }
        }
    }

    private fun stopSamplersLocked() {
        metricSampler?.stop(); metricSampler = null
        fpsSampler?.stop(); fpsSampler = null
        hangDetector?.stop(); hangDetector = null
    }

    // ---------------------------------------------------------------------
    // Analytics
    // ---------------------------------------------------------------------

    @JvmStatic
    @JvmOverloads
    fun analytics(event: String, params: Map<String, Any?> = emptyMap(), screen: String? = null) {
        if (!isEnabled) return
        val payload = mutableMapOf<String, Any?>(
            "kind" to "analytics",
            "t" to now(),
            "event" to event,
            "params" to Envelope.sanitizeMap(params),
            "screen" to screen,
        )
        emit("SimConsole.analytics", Level.Info, payload)
    }

    @JvmStatic
    @JvmOverloads
    fun screen(name: String, params: Map<String, Any?> = emptyMap()) {
        if (!isEnabled) return
        val payload = mapOf<String, Any?>(
            "kind" to "screen",
            "t" to now(),
            "screen" to name,
            "params" to Envelope.sanitizeMap(params),
        )
        emit("SimConsole.analytics", Level.Info, payload)
    }

    // ---------------------------------------------------------------------
    // Network
    //
    // Three-stage emission per request — mirrors iOS Sources/SimConsole.swift:
    //   1. net.request   — method + URL + headers + body, when the request fires
    //   2. net.response  — status + duration + headers + body, when it returns
    //                      (or net.error if the call failed)
    //   3. net.header / net.body — one envelope per header and one for the body,
    //                      so a single bloated header doesn't truncate the row
    //
    // The panel pairs everything by `id`. Bodies are clipped to maxBodyChars.
    // ---------------------------------------------------------------------

    @JvmStatic
    @JvmOverloads
    fun networkRequest(
        id: String,
        method: String,
        url: String,
        headers: Map<String, String> = emptyMap(),
        body: String? = null,
        mocked: Boolean = false,
    ) {
        if (!isEnabled) return
        val payload = mutableMapOf<String, Any?>(
            "kind" to "net.request",
            "t" to now(),
            "id" to id,
            "method" to method,
            "url" to url,
        )
        if (mocked) payload["mocked"] = true
        emit("SimConsole.network", Level.Info, payload)
        emitHeaders(id, "request", headers)
        emitBody(id, "request", body)
    }

    @JvmStatic
    @JvmOverloads
    fun networkResponse(
        id: String,
        status: Int,
        durationMs: Int,
        headers: Map<String, String> = emptyMap(),
        body: String? = null,
        byteSize: Int? = null,
        mocked: Boolean = false,
    ) {
        if (!isEnabled) return
        val payload = mutableMapOf<String, Any?>(
            "kind" to "net.response",
            "t" to now(),
            "id" to id,
            "status" to status,
            "duration_ms" to durationMs,
        )
        if (byteSize != null) payload["byte_size"] = byteSize
        if (mocked) payload["mocked"] = true
        emit("SimConsole.network", Level.Info, payload)
        emitHeaders(id, "response", headers)
        emitBody(id, "response", body)
    }

    @JvmStatic
    fun networkError(id: String, durationMs: Int, error: String) {
        if (!isEnabled) return
        val payload = mapOf<String, Any?>(
            "kind" to "net.error",
            "t" to now(),
            "id" to id,
            "duration_ms" to durationMs,
            "error" to error,
        )
        emit("SimConsole.network", Level.Warn, payload)
    }

    private fun emitHeaders(id: String, direction: String, headers: Map<String, String>) {
        if (headers.isEmpty()) return
        for ((name, value) in headers) {
            val payload = mapOf<String, Any?>(
                "kind" to "net.header",
                "t" to now(),
                "id" to id,
                "direction" to direction,
                "name" to name,
                "value" to clip(value).orEmpty(),
            )
            emit("SimConsole.network", Level.Info, payload)
        }
    }

    private fun emitBody(id: String, direction: String, body: String?) {
        if (body.isNullOrEmpty()) return
        val payload = mapOf<String, Any?>(
            "kind" to "net.body",
            "t" to now(),
            "id" to id,
            "direction" to direction,
            "body" to clip(body),
        )
        emit("SimConsole.network", Level.Info, payload)
    }

    private fun clip(s: String?): String? {
        if (s.isNullOrEmpty()) return s
        val max = maxBodyChars
        if (s.length <= max) return s
        return s.substring(0, max) + "…[+${s.length - max} chars]"
    }

    // ---------------------------------------------------------------------
    // Generic log
    // ---------------------------------------------------------------------

    @JvmStatic
    @JvmOverloads
    fun log(message: String, level: Level = Level.Info, fields: Map<String, Any?> = emptyMap()) {
        if (!isEnabled) return
        val payload = mapOf<String, Any?>(
            "kind" to "log",
            "t" to now(),
            "level" to level.toWire(),
            "msg" to message,
            "fields" to Envelope.sanitizeMap(fields),
        )
        emit("SimConsole.event", level, payload)
    }

    // ---------------------------------------------------------------------
    // Internal emission path
    // ---------------------------------------------------------------------

    /** Used by [Metric] (separate file, same package). Pinned to `Level.Info`. */
    internal fun emitMetric(payload: Map<String, Any?>) {
        emit("SimConsole.metric", Level.Info, payload)
    }

    private fun emit(tag: String, level: Level, payload: Map<String, Any?>) {
        val json = Envelope.encode(payload)
        if (!LogChunker.isOversized(json)) {
            sink.emit(tag, level, json)
            return
        }
        val id = UUID.randomUUID().toString()
        for (chunk in LogChunker.chunk(id, json)) {
            sink.emit("$tag.chunk", level, chunk)
        }
    }

    private fun now(): Double = System.currentTimeMillis() / 1000.0

    private fun Level.toWire(): String = when (this) {
        Level.Debug -> "debug"
        Level.Info -> "info"
        Level.Warn -> "warn"
        Level.Error -> "error"
    }

    /** Default sink — writes to `android.util.Log`. */
    private object LogcatSink : Sink {
        override fun emit(tag: String, level: Level, payload: String) {
            when (level) {
                Level.Debug -> Log.d(tag, payload)
                Level.Info -> Log.i(tag, payload)
                Level.Warn -> Log.w(tag, payload)
                Level.Error -> Log.e(tag, payload)
            }
        }
    }
}
